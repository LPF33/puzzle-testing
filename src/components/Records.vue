<template>
    <div>
        <h1>Records</h1>
        <div v-for="(record, index) of records" :key="index">
            {{ index + 1 }}. - {{ record[1] }} min
        </div>
    </div>
</template>
<script>
export default {
    name: "Records",
    props: ["puzzleId"],
    data() {
        return {
            records: [],
        };
    },
    created() {
        this.getRecords();
    },
    methods: {
        getRecords() {
            try {
                let storedRecords = JSON.parse(localStorage.getItem("records"));

                const sortedRecords = storedRecords
                    .filter((item) => item[0] == this.puzzleId)
                    .sort((a, b) => a[1] - b[1])
                    .map((item) => {
                        const milli = new Date(item[1]);
                        item[1] = milli.toISOString().slice(14, 19);
                        return item;
                    });
                this.records = sortedRecords;
            } catch (error) {
                this.records = [];
            }
        },
    },
};
</script>
<style scoped>
div,
h1 {
    text-shadow: 1px 1px 2px black;
}
h1 {
    border-bottom: 1px dashed black;
    margin-bottom: 10px;
}
</style>
