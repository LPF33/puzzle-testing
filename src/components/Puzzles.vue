<template>
    <div id="puzzles-component">
        <h1>Select a puzzle</h1>
        <div v-for="puzzle in puzzles" :key="puzzle.id" class="puzzle">
            <button @click="selectPuzzle(puzzle)" class="play-button">
                <div>
                    <img :src="require(`../assets/${puzzle.image}`)" />
                    <h2>{{ puzzle.title }}</h2>
                </div>
            </button>
            <Records :puzzleId="puzzle.id" />
        </div>
    </div>
</template>

<script>
import Records from "./Records";
export default {
    name: "Puzzles",
    components: {
        Records,
    },
    data() {
        return {
            puzzles: [
                { id: "owl", image: "owl.jpg", title: "Owl" },
                { id: "eggcozy", image: "eggcozy.jpg", title: "Egg cozy" },
            ],
        };
    },
    methods: {
        selectPuzzle(p) {
            this.$emit("play", p);
        },
    },
};
</script>

<style scoped>
#puzzles-component {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex-grow: 1;
}

.puzzle {
    position: relative;
    display: flex;
    max-width: 100vw;
    flex-wrap: wrap;
    justify-content: space-evenly;
    align-items: center;
}

.puzzle img {
    width: 300px;
    max-width: 100vw;
    border-radius: 5px;
}

.play-button {
    width: 800px;
    max-width: 100vw;
    height: 100%;
    background-color: #87a7b3;
    cursor: pointer;
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
}

.play-button h2 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
}

h1 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
    background-color: #e1f1dd;
    padding: 10px 0;
    margin-bottom: 10px;
}
</style>
