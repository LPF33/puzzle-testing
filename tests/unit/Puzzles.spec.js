import { mount } from "@vue/test-utils";
import Puzzles from "@/components/Puzzles.vue";
import Records from "@/components/Records.vue";

test("mount Puzzles component", () => {
    const wrapper = mount(Puzzles);
    expect(wrapper.html()).toMatchSnapshot();
});

test("mount with different data", () => {
    const wrapper = mount(Puzzles, {
        data() {
            return {
                puzzles: [
                    { id: "horse", image: "owl.jpg", title: "Horse" },
                    { id: "egg", image: "eggcozy.jpg", title: "Egg" },
                    { id: "apple", image: "eggcozy.jpg", title: "Apple" },
                ],
            };
        },
    });

    const puzzleClass = wrapper.findAll(".puzzle");
    expect(puzzleClass).toHaveLength(3);

    const h2s = wrapper.findAll("h2");
    expect(h2s).toHaveLength(3);
    expect(h2s[0].html()).toBe("<h2>Horse</h2>");
    expect(h2s[1].html()).toBe("<h2>Egg</h2>");
    expect(h2s[2].html()).toBe("<h2>Apple</h2>");

    const buttons = wrapper.findAll(".play-button");
    expect(buttons).toHaveLength(3);
});

test("when mounted every puzzle has Records component", () => {
    const wrapper = mount(Puzzles);
    const records = wrapper.findAllComponents(Records);
    expect(records).toHaveLength(2);
    expect(records[0].props()).toEqual({ puzzleId: "owl" });
    expect(records[1].props()).toEqual({ puzzleId: "eggcozy" });
});

test("when first play button clicked, event play emitted with first puzzle", async () => {
    const wrapper = mount(Puzzles);
    const buttons = wrapper.findAll(".play-button");
    expect(buttons).toHaveLength(2);
    await buttons[0].trigger("click");
    expect(wrapper.emitted()).toHaveProperty("play");
    expect(wrapper.emitted().play).toEqual([[wrapper.vm.puzzles[0]]]);
});

test("when second play button clicked, event play emitted with second puzzle", async () => {
    const wrapper = mount(Puzzles);
    const buttons = wrapper.findAll(".play-button");
    expect(buttons).toHaveLength(2);
    await buttons[1].trigger("click");
    expect(wrapper.emitted()).toHaveProperty("play");
    expect(wrapper.emitted().play).toEqual([[wrapper.vm.puzzles[1]]]);
});
