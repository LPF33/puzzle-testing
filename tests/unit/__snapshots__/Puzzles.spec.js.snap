// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount Puzzles component 1`] = `
<div id="puzzles-component">
  <h1>Select a puzzle</h1>
  <div class="puzzle"><button class="play-button">
      <div><img src="">
        <h2>Owl</h2>
      </div>
    </button>
    <div>
      <h1>Records</h1>
    </div>
  </div>
  <div class="puzzle"><button class="play-button">
      <div><img src="">
        <h2>Egg cozy</h2>
      </div>
    </button>
    <div>
      <h1>Records</h1>
    </div>
  </div>
</div>
`;
